Feature/tony/payment/bank-cd

TonyKim9401 opened the pull request:
# 해결하려는 문제가 무엇인가요?
* 은행 계좌를 사용한 사용자의 결제수단 등록 및 삭제

# 어떻게 해결했나요?
1. ### 결제 수단 2가지로 축소
- 은행 계좌 정보
- 카드 정보

2. ### 계좌 / 카드 등록 기능 구현중
#### 성공 케이스
- `사용자가 은행을 선택하여 계좌 결제 정보를 등록한다`
- `사용자가 카드 결제 정보를 등록한다`
- `계좌 정보 인증에 성공한다`
- `사용자가 등록한 계좌 정보를 조회한다`
- `사용자가 등록한 계좌 정보를 삭제한다`
- `사용자가 카드를 결제 정보로 등록한다`
- `(계좌, 계좌)사용자가 기존 기본 결제 정보가 있지만, 새로 기본 결제 정보를 저장 하려고 하면, 새로 등록되는 계좌가 기본 결제 정보로 등록된다`
- `(카드, 계좌)사용자가 기존 기본 결제 정보가 있지만, 새로 기본 결제 정보를 저장 하려고 하면, 새로 등록되는 계좌가 기본 결제 정보로 등록된다`
- `(카드, 카드)사용자가 기존 기본 결제 정보가 있지만, 새로 기본 결제 정보를 저장 하려고 하면, 새로 등록되는 계좌가 기본 결제 정보로 등록된다`
#### 실패 케이스
- `인증되지 않은 계좌는 등록할 수 없다`
- `기등록 계좌 정보면 등록 할 수 없다`
- `사용자가 기등록한 카드 정보면 등록 할 수 없다`
- `사용자가 새로 기본 결제 정보를 저장하려는데, 기존 기본 정보가 그대로 남아있다면 저장할 수 없다`

3. ### 결제 수단 인증 외부 API 구현 되지 않음
- 결제 수단 인증 외부 API 클래스의 경우 현재 무조건 true를 반환하도록 진행 중
## Attachment

JangHyeonJun2 committed:
add: testContainer

JangHyeonJun2 committed:
delete: KatjApplicationTests

TonyKim9401 committed:
update: 은행 이름 검증 기능 추가

결제수단 중 계좌 정보 추가시 어플케이션에서 사용 가능한 은행인지 검증

TonyKim9401 committed:
update: 은행 이름 검증 기능 추가

결제수단 중 계좌 정보 추가시 어플케이션에서 사용 가능한 은행인지 검증

TonyKim9401 committed:
update: ENUM 예외 메세지 추가

해당 값이 존재하지 않을시 400에러
인증되지 않은 결제 수단일시 500에서
에 대한 메세지 추가

TonyKim9401 committed:
update: ENUM 예외 메세지 추가

해당 값이 존재하지 않을시 400에러
인증되지 않은 결제 수단일시 500에서
에 대한 메세지 추가

TonyKim9401 committed:
update: 결제수단 내용 변경 메서드 추가

1.addPaymentMethodTo
사용자의 결제 수단으로 등록

2-1toResponsePaymentMethod
기본 결제수단으로 등록

2-2changeDefaultToFalse
기본 결제수단에서 해제

3.changeValidToTrue
결제수단 인증 후 인증 확인으로 변경

4.toResponsePaymentMethod
서비스에서 객체 반환시 Dto 로 바꾸는 컨버타

TonyKim9401 committed:
update: 결제수단 내용 변경 메서드 추가

1.addPaymentMethodTo
사용자의 결제 수단으로 등록

2-1toResponsePaymentMethod
기본 결제수단으로 등록

2-2changeDefaultToFalse
기본 결제수단에서 해제

3.changeValidToTrue
결제수단 인증 후 인증 확인으로 변경

4.toResponsePaymentMethod
서비스에서 객체 반환시 Dto 로 바꾸는 컨버타

TonyKim9401 committed:
update: 결제수단 계좌/카드로 변경

기존 신용 카드 / 체크 카드
-> 카드로 통일

TonyKim9401 committed:
update: 결제수단 계좌/카드로 변경

기존 신용 카드 / 체크 카드
-> 카드로 통일

TonyKim9401 committed:
update: 결제 수단 Controller 사용 dto 추가

1.RequestAddBankAccount
계좌 정보 추가시 사용 dto, 엔터티로 컨버터 기능

2.RequestAddCard
카드 정보 추가시 사용 dto, 엔터티로 컨버터 기능

3.ResponsePaymentMethod
조회 후 반환용 객체

TonyKim9401 committed:
update: 결제 수단 Controller 사용 dto 추가

1.RequestAddBankAccount
계좌 정보 추가시 사용 dto, 엔터티로 컨버터 기능

2.RequestAddCard
카드 정보 추가시 사용 dto, 엔터티로 컨버터 기능

3.ResponsePaymentMethod
조회 후 반환용 객체

TonyKim9401 committed:
update: 결제수단을 사용자에게 등록

addPaymentMethodToUser
추가한 결제 수단을 사용자에 등록 한다.

TonyKim9401 committed:
update: 결제수단을 사용자에게 등록

addPaymentMethodToUser
추가한 결제 수단을 사용자에 등록 한다.

TonyKim9401 committed:
add: 계좌 / 카드 외부 인증 API

외부 API가 연동되진 않았지만 추가 예정
무조건 인증된것으로 하고 true 반환

TonyKim9401 committed:
add: 계좌 / 카드 외부 인증 API

외부 API가 연동되진 않았지만 추가 예정
무조건 인증된것으로 하고 true 반환

TonyKim9401 committed:
update: 결제수단 Repository 수정

결제수단 Component 계층에서 사용 할 메서드 추가

즐겨찾기때와 마찬가지로 delete기능시 JPQL로 재작성 함

TonyKim9401 committed:
update: 결제수단 Repository 수정

결제수단 Component 계층에서 사용 할 메서드 추가

즐겨찾기때와 마찬가지로 delete기능시 JPQL로 재작성 함

TonyKim9401 committed:
add: 결제수단 Component계층 생성

PaymentMethodManager
결제수단 관련 생성, 삭제 기능

PaymentMethodReader
결제수단 관련 조회 기능

TonyKim9401 committed:
add: 결제수단 Component계층 생성

PaymentMethodManager
결제수단 관련 생성, 삭제 기능

PaymentMethodReader
결제수단 관련 조회 기능

TonyKim9401 committed:
add: PaymentMethodService 추가

은행 계좌 / 카드 크게 둘로 나누어 기능 구현중

TonyKim9401 committed:
add: PaymentMethodService 추가

은행 계좌 / 카드 크게 둘로 나누어 기능 구현중

TonyKim9401 committed:
add: PaymentMethodServiceTest 추가

TDD로 PaymentMethodService 구축중
현재 테스트 하나 빼고 모두 작동

*계좌 정보 인증에 실패 기능은 외부 API 연동후 가능하므로 임시 @Disabled 처리

TonyKim9401 committed:
add: PaymentMethodServiceTest 추가

TDD로 PaymentMethodService 구축중
현재 테스트 하나 빼고 모두 작동

*계좌 정보 인증에 실패 기능은 외부 API 연동후 가능하므로 임시 @Disabled 처리

JangHyeonJun2 commented on src/main/kotlin/com/hh2/katj/payment/model/entity/PaymentMethod.kt:
```diff
@@ -86,5 +86,36 @@ class PaymentMethod (
+                this.isDefault = false
+        }
+
+        fun 터() {
```
오타인거 같습니다.

TonyKim9401 replied:
제가 맥에서 인텔리제이를 잘 못 다루는 건지...
뭔가 버퍼 걸리듯이 타이핑한게 저런식으로 가있네요 잡아주셔서 감사해요 ㅠㅜ

JangHyeonJun2 commented on src/main/kotlin/com/hh2/katj/payment/repository/PaymentMethodRepository.kt:
```diff
@@ -1,9 +1,33 @@
+    @Query("SELECT pm FROM PaymentMethod pm WHERE pm.user.id = :userId AND pm.isDefault = true")
+    fun findAllByUserIdAndIsDefault(@Param("userId") userId: Long): PaymentMethod?
+
+    @Modifying
```
혹시 `@modifying` 어노테이션의 의미를 알 수 있을까요? 
주로 `Update, Insert, DELETE` 쿼리에서 쓰이고 벌크쪽으로 쓰이는걸로 알고 있는데, SELECT 문에서 쓰신 이유가 궁금합니다.

TonyKim9401 replied:
오 SELECT에 실수로 붙였었나보네요 감사합니다.
@Modifying 어노테이션을 붙이면 Spring Data JPA에서 메서드가 데이터 수정작업임을 인지하고 트랜잭션 적용 및 영속 컨텍스트를 좀 더 적절하게 사용하는걸로 알고있어서 붙이고 있습니다!

JangHyeonJun2 replied:
넵넵 :)

JangHyeonJun2 commented on src/main/kotlin/com/hh2/katj/payment/repository/PaymentMethodRepository.kt:
```diff
@@ -1,9 +1,33 @@
+
+    fun findAllByUserId(userId: Long): List<PaymentMethod>
+
+    @Modifying
```
단건 삭제인경우인거 같은데 여기서도 `@modifying` 어노테이션 사용하신 이유가 궁금합니다.

TonyKim9401 replied:
위에 내용으로 답변이 되었을까요 :)

JangHyeonJun2 commented on src/main/kotlin/com/hh2/katj/payment/service/PaymentMethodService.kt:
```diff
@@ -0,0 +1,177 @@
+        /**
+         * 요청 계좌 인증 유/무 확인
+         */
+        if (!request.isValid!!) {
```
이거는 사실 코드 성향인거 같은데, addBankAccount 파라미터에서 받는 request가 지금 paymentMethod 객체인거 같은데, request 보다는 paymentMehod 도 괜찮을거같습니다. 

또한 service 레이어에서 exception 호출도 괜찮겠지만 controller 단에서 검증이후 호출해도 괜찮지 않을까? 생각되네요!

TonyKim9401 replied:
request라고 적은 이유는 update, delete 등의 작업시 request.id로 db를 조회하여 payment객체를 실제로 쓸 때를 염두하여 request와 payment가 헷갈리지 않기 위해 명명했던 거구요.
파라미터에서 받는 객체를 request가 아닌 paymentMethod로 변경시, 위에 제가 예시로 든 수정,삭제 등 수행 전에 실제 객체 이름은 혹시 좋은 추천 있으시면 변경해도 좋다고 생각됩니다 ㅎㅎ 예시 주시면 적극 반영하도록 할께요!

아! 그리고 계좌 / 카드 인증 유/무에 대해서는1
1. 인증 API를 따로 만들어 인증 처리
2. 등록 API에 넘어오는 정보들은 무조건 인증이 완료됨

으로 생각하고 있어서요! 
그럼에도 불구하고 등록 서비스에 인증 유/무 확인이 있는건 2중 3중으로 검증을 처리하기 위한 안전장치로서 해두었습니다!
인증 API와 등록 API를 따로 두려고 생각하였기 때문에 등록 API 쪽에서는 간단하게 true, false만 가지고 확인하게 해두었구요!
다른 좋은 의견 있으시면 말씀 부탁드려요

JangHyeonJun2 commented on src/main/kotlin/com/hh2/katj/payment/service/PaymentMethodService.kt:
```diff
@@ -0,0 +1,177 @@
+        }
+    }
+
+    fun findOnePaymentMethod(userId: Long, paymentMethodId: Long): ResponsePaymentMethod {
```
paymentMethod 안에 유저 정보가 있는데, 파라미터에서 USER_ID와 같이 넘기는 이유가 있을까요!?

TonyKim9401 replied:
클라이언트에서 API 호출시 유효성 등이 검증된 userId를 외부에서 받아서 쓰는게 좋지 않을까 하여 분리해뒀는데 paymentMethod객체가 들고있는 유저 정보가 사용불가가 되거나 할 경우를 생각해서 해뒀는데. 빼는게 좋을까요?

JangHyeonJun2 replied:
흠.. 기훈님이 말씀하신것처럼  컨트롤러단에서 받는다면 그걸 같이 사용하는게 좋을거같네요 :)

JangHyeonJun2 commented on src/main/kotlin/com/hh2/katj/payment/service/PaymentMethodValidationApi.kt:
```diff
@@ -0,0 +1,16 @@
+
+@Component
+class PaymentMethodValidationApi {
+    fun bankAccountValidation(paymentMethod: PaymentMethod): Boolean{
```
이것도 코스 스타일이긴 하겠지만, 개인적으로 리턴값으로 boolean주는 것보다 검증을 통과하면 객체를 넘긴다는지?? 
요런것도 생각해볼 수 있겠네요!

TonyKim9401 replied:
아 이건 외부 API 연동이 어떻게 될지 몰라 임시로 Boolean값으로 무조건 인증된걸로 넘기려고 해뒀습니다!
말씀 주신 내용 참고해서 외부 검증 API 사용시 사용해보도록 할께요 👍

JangHyeonJun2 replied:
넵넵 !!

JangHyeonJun2 commented on src/test/kotlin/com/hh2/katj/payment/service/PaymentMethodServiceTest.kt:
```diff
@@ -0,0 +1,568 @@
+    }
+
+    @Test
+    fun `사용자가 은행을 선택하여 계좌 결제 정보를 등록한다`() {
```
💯

TonyKim9401 replied:
코드리뷰 고생하셨습니다 감사합니다!

JangHyeonJun2 committed:
update: testContainer 분리

JangHyeonJun2 committed:
Merge branch 'dev' into feature/lil/addTestContainer

# Conflicts:
#	build.gradle.kts
#	src/main/kotlin/com/hh2/katj/favorite/repository/FavoriteRepository.kt
#	src/main/kotlin/com/hh2/katj/favorite/service/FavoriteService.kt

JangHyeonJun2 committed:
update: 불필요한 의존성 삭제

TonyKim9401 committed:
update: 현준님 코드리뷰 반영하여 코드 수정

TonyKim9401 committed:
update: 현준님 코드리뷰 반영하여 코드 수정

JangHyeonJun2 committed:
Merge pull request #82 from KATJ-HH2/feature/lil/addTestContainer

add: testContainer

TonyKim9401 committed:
update: FavoriteManager의 기능 FavoriteReader로 분리

1.FavoriteReader
기존 메서드들의 정리가 되어있지 않아 정리

2.FavoriteManager
즐겨찾기 도메인 중심의 조회 기능 Reader로 일부 옮김

3.FavoriteService
분리에 맞추어 리팩토링

TonyKim9401 committed:
update: FavoriteManager의 기능 FavoriteReader로 분리

1.FavoriteReader
기존 메서드들의 정리가 되어있지 않아 정리

2.FavoriteManager
즐겨찾기 도메인 중심의 조회 기능 Reader로 일부 옮김

3.FavoriteService
분리에 맞추어 리팩토링

TonyKim9401 committed:
Merge remote-tracking branch 'origin/feature/tony/payment/bank-CD' into feature/tony/payment/bank-CD

TonyKim9401 committed:
update: 카드 유효기간 검증 기능 추가

카드 유효기간이
1.유효하면 true 반환
2.유효하지 않으면 INVALID_PAYMENT_METHOD 오류 발생

TonyKim9401 committed:
update: 카드 유효기간 검증 기능 테스트

카드 유효기간이
1.유효하면 true 반환
2.유효하지 않으면 INVALID_PAYMENT_METHOD 오류 발생

JangHyeonJun2 reviewed (approved):
고생하셨습니다.
